### PR TITLE
reduce GC load

### PR DESCRIPTION
**Pull request by florianl**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Looping over an array to append something can put high preasure on the garbage collector and cause memory relocations. If the destination is properly allocated in advance, it reduces the work that is needed to be done by GC.

This is not a enduser visible change so I did not add something for the release notes.

### CODE REVIEW

**Review by michi-covalent — APPROVED** [review]

File: pkg/kvstore/store/store.go [hunk 1]
```diff
@@ -297,7 +297,7 @@ func (s *SharedStore) syncLocalKeys(ctx context.Context) error {
 	// Create a copy of all local keys so we can unlock and sync to kvstore
 	// without holding the lock
 	s.mutex.RLock()
-	keys := []LocalKey{}
+	keys := make([]LocalKey, len(s.localKeys))
 	for _, key := range s.localKeys {
 		keys = append(keys, key)
 	}
```
Comment (rolinh) [inline pkg/kvstore/store/store.go]: This doesn't do what you want as you'll end up with a slice that is `2*len(s.localKeys)` large with the first half of values being the zero value of `LocalKey` after the `append` loop. What you want to do instead is to preallocate a slice of _capacity_ `len(s.localKeys)`, ie: `keys := make([]LocalKey, 0, len(s.localKeys))`.

File: pkg/option/option.go [hunk 2]
```diff
@@ -326,7 +326,7 @@ func (o *IntOptions) GetFmtList() string {
 	txt := ""
 
 	o.optsMU.RLock()
-	opts := []string{}
+	opts := make([]string, len(o.Opts))
```
Comment (rolinh) [inline pkg/option/option.go]: ditto

File: pkg/option/option.go [hunk 3]
```diff
@@ -349,7 +349,7 @@ func (o *IntOptions) Dump() {
 	}
 
 	o.optsMU.RLock()
-	opts := []string{}
+	opts := make([]string, len(o.Opts))
```
Comment (rolinh) [inline pkg/option/option.go]: ditto

**Review by rolinh — CHANGES_REQUESTED** [review]

**Review by rolinh — APPROVED** [review]

**Review by sayboras — COMMENTED** [review]
Thanks, just curious if there is any way to detect similar occurrence in the codebase with linter.

Goland code inspection is showing ~150+ similar but weak occurrences

image

**Review by aanm — APPROVED** [review]

**Comment by florianl** [comment]
> Thanks, just curious if there is any way to detect similar occurrence in the codebase with linter.

Finding such cases with linter rules might be possible. In my case I was looking at a flamegraph produced by a continous profiler and looked into cases where `runtime.growslice` happens often.

**Review by sayboras — APPROVED** [review]

**Comment by florianl** [comment]
Is there something left I can do?

**Review by jibi — APPROVED** [review]

**Comment by florianl** [comment]
sorry for the noise. but Is there something left I can do?

**Comment by rolinh** [comment]
@florianl Nope. All required reviews are in. Let's kick the CI.

**Comment by florianl** [comment]
Checked some of the failing tests. The reason of the failure seems to be the same for all. But I don't think this is related to the changes in this PR.

| Failed test | Fail reason|
|-----------|-------------|
| Cilium L4LB XDP  | The action has timed out.  |
| AWS_CNI |The action has timed out.|
| EKS |The action has timed out.|
| GKE |The action has timed out.|
| External workloads |The action has timed out.|
| Multicluster |The action has timed out.|

**Comment by pchaigno** [comment]
> Checked some of the failing tests. The reason of the failure seems to be the same for all. But I don't think this is related to the changes in this PR.

Yes and no. They timed out waiting for the CI images to be ready. The images are not there because they were built more than 7 days ago and thus have been recycled in the meantime. I've restarted the builds and I'll restart the end-to-end tests once the images are ready.

**Review by kaworu — APPROVED** [review]

**Pull request merged**